### PR TITLE
python312Packages.wxpython: fix build

### LINKED ISSUE #326764: Build failure: "wxpython-4.2.1 not supported for interpreter python3.12"
### Steps To Reproduce

1. `nix-shell -p python3Packages.wxpython`

### Build log

```
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'shell'
         whose name attribute is located at /nix/store/98mgnj8fy0vx0l7yq1lzls93d5h6zhmv-nixos/nixos/pkgs/stdenv/generic/make-derivation.nix:332:7

       … while evaluating attribute 'buildInputs' of derivation 'shell'

         at /nix/store/98mgnj8fy0vx0l7yq1lzls93d5h6zhmv-nixos/nixos/pkgs/stdenv/generic/make-derivation.nix:379:7:

          378|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          379|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          380|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       error: wxpython-4.2.1 not supported for interpreter python3.12
```

### Notify maintainers

@mweinelt

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.9.7, NixOS, 24.11 (Vicuna), 24.11pre647193.9f4128e00b0a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.4`
 - channels(root): `"nixos, nixos-hardware"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by wegank**
## Description of changes

ngi-nix/ngipkgs#311 and #325726 both pin Python to version 3.11 for certain packages, as `python312Packages.wxpython` is disabled in Nixpkgs, whose reason is probably that wxPython depends on distutils.

What confuses me is that there are wheels for Python 3.12 on PyPI (see https://pypi.org/project/wxPython/#files), and there are no active issues about Python 3.12 support or the use of distutils in the wxPython issue tracker.

So there should be a way to build `python312Packages.wxpython`, right?

Capture d’écran 2024-07-10 à 18 25 39

Fixes #326764

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Noodlez1232** [comment]
IFAIK, wxPython for Pyhton 3.12 is in beta until this next release, meaning it's not fully supported. I think that's why it's disabled.

**Comment by Noodlez1232** [comment]
What I mean by that is even though it's included in a stable release, its support for 3.12 is beta.

**Comment by D3vil0p3r** [comment]
Will it fix also the error `error: wxpython-4.2.1 not supported for interpreter python3.12`?

**Review by Aleksanaa — APPROVED** [review]

File: pkgs/development/python-modules/wxpython/4.2.nix [hunk 1]
```diff
@@ -61,6 +67,15 @@ buildPythonPackage rec {
     })
   ];
 
+  postPatch = lib.optionalString (pythonAtLeast "3.12") ''
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxpython/4.2.nix:70]:
```suggestion
  postPatch = ''
```
Please do this unconditional so that we spot issues easier

File: pkgs/development/python-modules/wxpython/4.2.nix [hunk 2]
```diff
@@ -40,11 +41,16 @@
   six,
 }:
 
+let
+  waf_2_0_25 = fetchurl {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wxpython/4.2.nix:45]: Is this this waf https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/wa/waf/package.nix ? Can we just use that instead?

Comment (wegank) [reply pkgs/development/python-modules/wxpython/4.2.nix:45]: Yes, but 2.1.2 is too recent, wxPython doesn't build with it.

**Comment by StarGate01** [comment]
Once this package is stable, compiled and merged for 3.12, we should unpin KiCAD (#325726) .

**Comment by raidenovich** [comment]
> What I mean by that is even though it's included in a stable release, its support for 3.12 is beta.

I am curious to know about this aswell.

File: pkgs/development/python-modules/wxpython/4.2.nix (hunk 2 above)
Comment (jfly) [reply pkgs/development/python-modules/wxpython/4.2.nix:45]: @wegank are you sure that wxPython doesn't build with waf 2.1.2? I just tried (see [this branch](https://github.com/jfly/nixpkgs/tree/wxpython-fix)), and `nix build .#python3.pkgs.wxpython` seems to work for me.

Is it possible you meant to say that wxPython doesn't build with 2.1.1? At the time of writing, this PR (#326142)'s commit is 9dc9813a8d49817a36bd0d11ff57417c469102aa, whose parent is 7c1d59ff54058cf7b36ef0aea0aad7a0870a17f7, which [has waf 2.1.1](https://github.com/jfly/nixpkgs/blob/7c1d59ff54058cf7b36ef0aea0aad7a0870a17f7/pkgs/by-name/wa/waf/package.nix#L13), so I'm a little surprised to see you mention waf 2.1.2 here. I tried building wxPython with waf 2.1.1, and it pretty quickly errors out with a conflict in argparse about a `--python` option, which was reported upstream [here](https://gitlab.com/ita1024/waf/-/merge_requests/2371), and fixed in [this commit](https://gitlab.com/ita1024/waf/-/commit/bba5047716ea26d93f501a30a4c6aa0ce8c1642a), which first landed in waf 2.1.2.

tl;dr: nixpkgs now has [waf 2.1.2](https://github.com/NixOS/nixpkgs/pull/328237). If you rebase this PR, I am pretty sure you'll find that you can follow @SuperSandro2000's advice and just use `waf`.

Tangentially related, I found a (IMO) cleaner mechanism for overriding waf, which you can see in my branch: https://github.com/NixOS/nixpkgs/compare/master...jfly:nixpkgs:wxpython-fix. Feel free to steal it. Also happy to make a PR out of my branch instead you prefer.

File: pkgs/development/python-modules/wxpython/4.2.nix [hunk 3]
```diff
@@ -61,6 +67,15 @@ buildPythonPackage rec {
     })
   ];
 
+  postPatch = ''
+    cp ${waf_2_0_25} bin/waf-2.0.25
```
Comment (jfly) [inline pkgs/development/python-modules/wxpython/4.2.nix:71]: Suggestion: Link [the issue you filed upstream](https://github.com/wxWidgets/Phoenix/issues/2575) here for discoverability.

File: pkgs/development/python-modules/wxpython/4.2.nix [hunk 4]
```diff
@@ -61,6 +60,13 @@ buildPythonPackage rec {
     })
   ];
 
+  # https://github.com/wxWidgets/Phoenix/issues/2575
+  postPatch = ''
+    ln -s ${lib.getExe buildPackages.waf} bin/waf
```
Comment (jfly) [inline pkgs/development/python-modules/wxpython/4.2.nix:65]: This is the first time I've ever seen `buildPackages`. Can you teach me what it's for (or more importantly, why we're using it here)? Why is this preferable to adding `waf` to the top? Is this a trick to reach "above/past" the python package set into the top level package set?

**Review by jfly — APPROVED** [review]
:shipit: One question just for me to learn more about nixpkgs.

File: pkgs/development/python-modules/wxpython/4.2.nix (hunk 4 above)
Comment (wegank) [reply pkgs/development/python-modules/wxpython/4.2.nix:65]: This ensures that `waf` is native to the build system (so that it can be run during the build phase), and not to the host system, in a cross-compilation context.

Comment (jfly) [reply pkgs/development/python-modules/wxpython/4.2.nix:65]: (I'm trying to mark this discussion resolved, but I cannot find the button...)
